### PR TITLE
added GroupedConvolutions and ChannelShuffle

### PR DESCRIPTION
**Pull request by gartangh**

### CODE REVIEW

File: Manifest.toml [hunk 1]
```diff
@@ -8,15 +8,15 @@ version = "0.5.0"
 
 [[AbstractTrees]]
 deps = ["Markdown"]
-git-tree-sha1 = "8201f932428d25a2e2903300764515754847d87d"
```
Comment (maleadt) [inline Manifest.toml]: Are the Manifest changes required? If not, leave them out.

File: src/layers/basic.jl [hunk 2]
```diff
@@ -214,7 +214,7 @@ size(sm(x)) == (5, 5, 11, 10)
 """
 struct SkipConnection
   layers
-  connection  #user can pass arbitrary connections here, such as (a,b) -> a + b
+  connection  # user can pass arbitrary connections here, such as (a,b) -> a + b
```
Comment (maleadt) [inline src/layers/basic.jl]: needless whitespace change

File: test/layers/basic.jl [hunk 3]
```diff
@@ -28,7 +28,7 @@ import Flux: activations
   end
 
   @testset "Dense" begin
-    @test  length(Dense(10, 5)(randn(10))) == 5
```
Comment (maleadt) [inline test/layers/basic.jl]: same

**Comment by maleadt** [comment]
I guess this will need to use cuDNN's support for grouped convolutions to run efficiently on the GPU? Did you write this code with that in mind? Also, @contributors, would that need an API in NNlib? We've kind-of been moving away from Flux using cuDNN internals.

File: Manifest.toml (hunk 1 above)
Comment (gartangh) [reply Manifest.toml]: I undid the manifest changes [in this commit](https://github.com/FluxML/Flux.jl/pull/1033/commits/0feabec499fa6cc10cab452240e7b29ac4b24671).

File: src/layers/basic.jl (hunk 2 above)
Comment (gartangh) [reply src/layers/basic.jl]: I undid the whitespace change [in this commit](https://github.com/FluxML/Flux.jl/pull/1033/commits/0feabec499fa6cc10cab452240e7b29ac4b24671).

File: test/layers/basic.jl (hunk 3 above)
Comment (gartangh) [reply test/layers/basic.jl]: I undid the whitespace change [in this commit](https://github.com/FluxML/Flux.jl/pull/1033/commits/0feabec499fa6cc10cab452240e7b29ac4b24671).

**Comment by gartangh** [comment]
@maleadt , I believe the [cuDNN API for grouped convolutions](https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#grouped-convolutions), can only run more efficiently when all `paths` consist of the same convolution and `split=true`. This is the case for Xception and ResNeXt.
This implementation also provides support for each path acting on the complete input (with `split=false`) and paths being different among each other (so different operations on different parts of the input). This is used in Inception variants.

**Comment by maleadt** [comment]
> the [cuDNN API for grouped convolutions](https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#grouped-convolutions), can only run more efficiently when all `paths` consist of the same convolution and `split=true`

Right, so your current design does not allow for use of cuDNN. If you, e.g. move the split field into the typevar you could dispatch on that (and T being a homogeneous NTuple, I presume).

**Comment by dhairyagandhi96** [comment]
Could you add some precedent here for these layers and their general popularity? PyTorch does this via a `groups` kwarg, which seems cleaner since NNlib can already support that

**Pull request closed without merging**